### ISSUE TITLE
Hang with STM & two scopes

### ISSUE THREAD

**Issue by ocharles** [awkward-squad/ki#33]
The following program hangs indefinitely:

```haskell
{-# LANGUAGE BlockArguments #-}
{-# LANGUAGE DeriveAnyClass #-}

module Main where

import Control.Monad
import Control.Exception
import Control.Concurrent
import Control.Concurrent.STM
import Ki

data Bang = Bang
  deriving (Exception, Show)

main :: IO ()
main = 
  scoped \scope1 -> do
    explodingThread <- fork scope1 do
      threadDelay 1_000_000
      throwIO Bang

    scoped \ki -> do
      c <- newTChanIO

      threadId <- fork ki do
        forever do
          atomically do
            readTChan c

      putStrLn "Waiting"
      threadDelay 10_000_000
```

I would expect this to be fine, and die with the `Bang` exception. If we use a single scope, that's exactly what happens, so it seems to be some kind of interation between two scopes.

**Comment by ocharles** [comment]
We also tried just using `threadDelay` and this code _does_ work, so it seems to be something specific to the use of STM:

```haskell
{-# LANGUAGE BlockArguments #-}
{-# LANGUAGE DeriveAnyClass #-}

module Main where

import Control.Monad
import Control.Exception
import Control.Concurrent
import Control.Concurrent.STM
import Ki

data Bang = Bang
  deriving (Exception, Show)

main :: IO ()
main = 
  scoped \scope1 -> do
    explodingThread <- fork scope1 do
      threadDelay 1_000_000
      throwIO Bang

    scoped \ki -> do
      c <- newTChanIO

      threadId <- fork ki do
        threadDelay 5_000_000

      putStrLn "Waiting"
      threadDelay 10_000_000
```

**Comment by ocharles** [comment]
One more finding:

If I do

```haskell
      threadId <- fork ki do
        print =<< try @SomeException do
          forever do
            atomically do
              readTChan c
          pure ()
```

I get

```
Left thhuh.hs: thread blocked indefinitely in an STM transaction
read blocked indefinitely in an ⏎
```

So I think it does get the exception and something else is borked

**Comment by mitchellwrosen** [comment]
Oh dear, thanks for the bug report.

**Comment by mitchellwrosen** [comment]
I think the bug is here:

https://github.com/awkward-squad/ki/blob/main/ki/src/Ki/Internal/Scope.hs#L169-L173

The intention is to avoid bothering to throw `ScopeClosing` exception right back to the child thread that's just propagated something to us (the parent).

But the child thread identifier here isn't globally unique: for each scope, it starts at 0 and counts up. So, we can easily get our wires crossed and avoid propagating an exception down to a child that should receive it.

I can see a few options:

1. Just delete this line of code; it's fine to `throwTo` to a dead `ThreadId`. I'm not sure when GHC recycles `ThreadId` but surely it is not really possible to kill the wrong thread here. I'd assume `ThreadId` will wrap around after 2^64 ids have been assigned.
2. Add a unique "scope id" that we can include in various places to know what scope we're talking about. In this case, the exception that a child throws to its parent can include the scope id that it was created in. This will let us accurately elect not to throw an exception back to the child thread that threw to us.

**Comment by mitchellwrosen** [comment]
This is fixed in `ki-1.0.1.2`. We went with option (1); I think the motivation was rather poor for this "feature".

Kind of an embarrassing bug! 😓 But thank you again for reporting!

**Comment by ocharles** [comment]
Amazing, thank you so much! And don't feel bad - most bugs are like this! We'll give this a bash soon. Fortunately we only hit this bug when a bunch of other stuff has crashed, so it's quite - if you'll forgive the pun - exceptional 😉
